set_bucket_tags/app.py

```diff
@@ -0,0 +1,135 @@
+  resource_properties = event.get('ResourceProperties')
+  bucket_name = resource_properties.get('BucketName')
+  if not bucket_name:
+    raise ValueError(MISSING_BUCKET_NAME_ERROR_MESSAGE)
+  return bucket_name
+
+
+def get_bucket_tags(bucket_name):
+  '''Look up the bucket tags'''
+  client = get_s3_client()
+  response = client.get_bucket_tagging(Bucket=bucket_name)
+  log.debug(f'S3 bucket tags response: {response}')
+  tags = response.get('TagSet')
+  if not tags or len(tags) == 0:
+    raise Exception(f'No tags returned, received: {response}')
+  return tags
+
+
+def get_principal_arn(tags):
+  '''Find the value of the principal arn among the bucket tags'''
+  principal_arn_tag = 'aws:servicecatalog:provisioningPrincipalArn'
+  for tag in tags:
+    if tag.get('Key') == principal_arn_tag:
+      principal_arn_value = tag.get('Value')
+      return principal_arn_value
+  else:
+    raise ValueError('Could not derive a provisioningPrincipalArn from tags')
+
+
+def get_synapse_user_name(principal_arn):
+  '''Use the synapse id embedded in the principal arn to look up the synapse user name'''
+  synapse_id = principal_arn.split('/')[-1]
+  if not synapse_id.isdigit():
+    error_msg = (f'The synapse_id {synapse_id} derived from the principal_arn'
+      f'{principal_arn} is in an unexpected format')
+    raise ValueError(error_msg)
+  synapse_url = f'https://repo-prod.prod.sagebase.org/repo/v1/userProfile/{synapse_id}'
+  response = requests.get(synapse_url)
+  response.raise_for_status()
+  user_profile = response.json()
+  log.debug(f'Synapse user profile response: {user_profile}')
+  user_name = user_profile.get('userName')
+  return user_name
+
+
+def add_owner_email_tag(tags, synapse_username):
+  '''Add an OwnerEmail tag for the synapse email based on synapse usename'''
+  synapse_email = f'{synapse_username}@synapse.org'
+  owner_email_tag = next((tag for tag in tags if tag['Key'] == 'OwnerEmail'), None)
+  if owner_email_tag:
+    owner_email_tag['Value'] = synapse_email
+  else:
+    new_owner_email_tag = { 'Key': 'OwnerEmail', 'Value': synapse_email}
+    tags.append(new_owner_email_tag)
+  return tags
+
+
+@helper.create
+@helper.update
+def create_or_update(event, context):
+  log.debug('Received event: ' + json.dumps(event, sort_keys=False))
+  log.info('Start SetBucketTags Lambda processing')
+  log.debug('Received event: ' + json.dumps(event, sort_keys=False))
+  bucket_name = get_bucket_name(event)
+  tags = get_bucket_tags(bucket_name)
+  principal_arn = get_principal_arn(tags)
+  synapse_username = get_synapse_user_name(principal_arn)
+  tags = add_owner_email_tag(tags, synapse_username)
+  client = get_s3_client()
+  tagging_response = client.put_bucket_tagging(
+    Bucket=bucket_name,
+    Tagging={ 'TagSet': tags }
+    )
+  log.debug(f'Tagging response: {tagging_response}')
+  if tagging_response.get('ResponseMetadata').get('HTTPStatusCode') == 204:
+    log.debug('got 204 response, all is fine')
+    return True
```

Comment:
you should return the resource id so CFN can remove and update it.

Reply:
Where do I get that? Is this really necessary? I could see how it might be if I were creating something that persists and has more complicated behavior, but this just sets tags.

In this example released by AWS, none of that complicated behavior happens: https://aws.amazon.com/blogs/infrastructure-and-automation/aws-cloudformation-custom-resource-creation-with-python-aws-lambda-and-crhelper/

Let's discuss on Slack.

Reply:
Removed send_response after all. It's not necessary with crhelper. Discussed this with Khai and he saw that no resource is being created so there's no id to return.